nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py

```diff
@@ -0,0 +1,327 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import math
+import pickle
+import random
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import numpy as np
+import torch
+import torch.utils.data as pt_data
+from hydra.utils import instantiate
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+from pytorch_lightning.utilities import rank_zero_only
+from sacrebleu import corpus_bleu
+from sacremoses import MosesDetokenizer
+
+from nemo.collections.common.losses import SmoothedCrossEntropyLoss
+from nemo.collections.common.parts import transformer_weights_init
+from nemo.collections.nlp.data import TranslationDataset
+from nemo.collections.nlp.models.enc_dec_nlp_model import EncDecNLPModel
+from nemo.collections.nlp.models.machine_translation.mt_enc_dec_config import MTEncDecModelConfig
+from nemo.collections.nlp.modules.common import TokenClassifier
+from nemo.collections.nlp.modules.common.transformer import BeamSearchSequenceGenerator, TransformerEmbedding
+from nemo.core.classes.common import typecheck
+from nemo.utils import logging
+
+__all__ = ['MTEncDecModel']
+
+
+class MTEncDecModel(EncDecNLPModel):
+    """
+    Encoder-decoder machine translation model.
+    """
+
+    def __init__(self, cfg: MTEncDecModelConfig, trainer: Trainer = None):
+
+        self.setup_enc_dec_tokenizers(cfg)
```

Comment:
Add 
```python
        cfg = model_utils.convert_model_config_to_dict_config(cfg)
        cfg = model_utils.maybe_update_config_version(cfg)
```

Call the following utility methods from ModelPT before passing cfg. Assuming that cfg is a dataclass or a dictconfig is not guaranteed unless it passes through ModelPT constructor, so these two methods will force cfg to be valid

https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/rnnt_bpe_models.py#L65-L67

Reply:
added

Reply:
@titu1994 do we need to call these at this level? can't we just call them at ModelPT level?

Reply:
ModelPT calls these two internally in it's constructor. However, this above line is being called before calling super() init method so ModelPT doesn't get the chance to convert cfg to a dictconfig